examples_tests/01.HelloWorld/main.cpp

```diff
@@ -286,13 +302,236 @@ int main()
+	}
+
+	VkCommandBuffer commandBuffers[SC_IMG_COUNT];
+
+	VkCommandBufferAllocateInfo allocateInfo = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO };
+	allocateInfo.commandPool = commandPool;
+	allocateInfo.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+	allocateInfo.commandBufferCount = SC_IMG_COUNT;
+
+	assert(vkAllocateCommandBuffers(vk_device, &allocateInfo, commandBuffers) == VK_SUCCESS);
+
+	VkClearValue clearColor = { 0.2f, 0.2f, 0.3f, 1.f };
+
+	// Record commands in commandBuffers here
+	for (uint32_t i = 0u; i < SC_IMG_COUNT; ++i)
+	{
+		VkCommandBufferBeginInfo beginInfo = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
+
+		assert(vkBeginCommandBuffer(commandBuffers[i], &beginInfo) == VK_SUCCESS);
+
+		VkRenderPassBeginInfo renderPassBeginInfo = { VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO };
+		renderPassBeginInfo.renderPass = vk_renderPass;
+		renderPassBeginInfo.framebuffer = vk_framebuffers[i];
+		renderPassBeginInfo.renderArea.offset = { 0, 0 };
+		renderPassBeginInfo.renderArea.extent = swapchainExtent;
+		renderPassBeginInfo.clearValueCount = 1u;
+		renderPassBeginInfo.pClearValues = &clearColor;
+
+		vkCmdBeginRenderPass(commandBuffers[i], &renderPassBeginInfo, VK_SUBPASS_CONTENTS_INLINE);
+
+		// Do nothing
+
+		vkCmdEndRenderPass(commandBuffers[i]);
+
+		assert(vkEndCommandBuffer(commandBuffers[i]) == VK_SUCCESS);
+	}
+
+	video::ISwapchain* rawPointerToSwapchain = swapchain.get();
+
+	uint32_t currentFrameIndex = 0u;
+	while (!windowShouldClose_Global)
+	{
+		video::IGPUSemaphore* acquireSemaphore_frame = acquireSemaphores[currentFrameIndex].get();
+		video::IGPUSemaphore* releaseSemaphore_frame = releaseSemaphores[currentFrameIndex].get();
+		video::IGPUFence* fence_frame = frameFences[currentFrameIndex].get();
+
+		assert(device->waitForFences(1u, &fence_frame, true, ~0ull) == video::IGPUFence::ES_SUCCESS);
+
+		uint32_t imageIndex;
+		swapchain->acquireNextImage(~0ull, acquireSemaphores[currentFrameIndex].get(), nullptr,
+			&imageIndex);
+
+		// At this stage the final color values are output from the pipeline
+		// Todo(achal): Not really sure why are waiting at this pipeline stage for
+		// acquiring the image to render
+		VkPipelineStageFlags pipelineStageFlags = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+
+		VkSubmitInfo submitInfo = { VK_STRUCTURE_TYPE_SUBMIT_INFO };
+		submitInfo.waitSemaphoreCount = 1u;
+		submitInfo.pWaitSemaphores = &vk_acquireSemaphores[currentFrameIndex];
+		submitInfo.pWaitDstStageMask = &pipelineStageFlags;
+		submitInfo.commandBufferCount = 1u;
+		submitInfo.pCommandBuffers = &commandBuffers[imageIndex];
+		submitInfo.signalSemaphoreCount = 1u;
+		submitInfo.pSignalSemaphores = &vk_releaseSemaphores[currentFrameIndex];
+
+		// Make sure you unsignal the fence before expecting vkQueueSubmit to signal it
+		// once it finishes its execution
+		device->resetFences(1u, &fence_frame);
+
+		VkResult result = vkQueueSubmit(vk_graphicsQueue, 1u, &submitInfo, vk_frameFences[currentFrameIndex]);

```

Comment:
not sure this is correct, putting the fence here, means that we only wait for the GPU to finish rendering to an image in order to recycle it for rendering, when in-fact we should be waiting for the GPU to finish presenting it.

Reply:
> when in-fact we should be waiting for the GPU to finish presenting it.

This should be handled by the semaphore passed to `vkAcquireNextImageKHR` i.e. that semaphore won't get signalled until the presentation engine is done presenting it. So as long as we're not doing anything to the image until the semaphore signals (which is the case here, since we're waiting on it in `vkQueueSubmit`) we should be fine.